plugin/pkg/auth/authenticator/token/oidc/oidc.go

```diff
@@ -113,9 +186,73 @@ func New(issuerURL, clientID, caFile, usernameClaim, groupsClaim string) (*OIDCA
 	// SyncProviderConfig will start a goroutine to periodically synchronize the provider config.
 	// The synchronization interval is set by the expiration length of the config, and has a mininum
 	// and maximum threshold.
-	stop := client.SyncProviderConfig(issuerURL)
+	stop := client.SyncProviderConfig(issuer)
+
+	authn := &OIDCAuthenticator{
+		clientConfig:     ccfg,
+		client:           client,
+		usernameClaim:    usernameClaim,
+		groupsClaim:      groupsClaim,
+		stopSyncProvider: stop,
+	}
+	authn.bearerAuthenticator = bearertoken.New(authn)
+	return authn, nil
+}
 
-	return &OIDCAuthenticator{ccfg, client, usernameClaim, groupsClaim, stop}, nil
+// SetHandlersAttached tells the authenticator that the OIDC related HTTP handlers are in use and have been attached.
+//
+// When this is set to true, requests to the following paths will not require
+// authentication:
+//
+// PathExchangeRefreshToken, PathAuthenticate, PathAuthCallback,
+// PathExchangeCode
+func (a *OIDCAuthenticator) SetHandlersAttached(attached bool) {
+	a.handlersAttached = attached
+}
+
+// NewOIDCHTTPHandler creates an http.Handler which provides additional endpoints useful for OIDC-related authentication.
+//
+// An OIDCAuthenticator created with a client-secret is necessary, or else an
+// error is returned.
+func (a *OIDCAuthenticator) NewOIDCHTTPHandler(apiHostPort string) (*OIDCHTTPHandler, error) {
+	apiServerURL := url.URL{
+		Host:   apiHostPort,
+		Scheme: "https",
+	}
+
+	oac, err := a.client.OAuthClient()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &oidcClient{
+		oidc: a.client,
+		oac:  oac,
+	}
+
+	if a.clientConfig.Credentials.Secret == "" {
+		return nil, errors.New("OIDCHTTPHandler requires a client config with a client secret")
+	}
+
+	return &OIDCHTTPHandler{
+		client:       client,
+		apiServerURL: apiServerURL,
+	}, nil
+}
+
+func NewOIDCHTTPHandler(o OIDCClient) *OIDCHTTPHandler {
+	return &OIDCHTTPHandler{
+		client: o,
+	}
+}
+
+func (a *OIDCAuthenticator) AuthenticateRequest(req *http.Request) (user.Info, bool, error) {
+	if a.handlersAttached {
```

Comment:
this makes me nervous... I'd rather see the top-level call to the auth filter decide whether auth is needed or not